### PR TITLE
MysqlMutext component fix same connection but difference database fail

### PR DESCRIPTION
**Pull request by kamarton**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #19316

### CODE REVIEW

File: framework/mutex/MysqlMutex.php [hunk 1]
```diff
@@ -35,6 +36,11 @@
  */
 class MysqlMutex extends DbMutex
 {
+    /**
+     * @var Expression|string|null prefix value. If null (by default) then connection's current database name is used.
```
Comment (samdark) [inline framework/mutex/MysqlMutex.php:40]: Need `@since` tag.

**Review by samdark — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

File: framework/mutex/MysqlMutex.php [hunk 2]
```diff
@@ -59,8 +69,8 @@ protected function acquireLock($name, $timeout = 0)
         return $this->db->useMaster(function ($db) use ($name, $timeout) {
             /** @var \yii\db\Connection $db */
             return (bool) $db->createCommand(
-                'SELECT GET_LOCK(:name, :timeout)',
-                [':name' => $this->hashLockName($name), ':timeout' => $timeout]
+                'SELECT GET_LOCK(CONCAT(:prefix, :name), :timeout)',
```
[the remark this answers is not in the document]
Comment (kamarton) [reply framework/mutex/MysqlMutex.php:72]: @rob006 How about CRC32?

`SELECT GET_LOCK(CONCAT(CRC32(:prefix), :name), :timeout)`

`CRC32` is supported by mysql and mariadb:
- https://dev.mysql.com/doc/refman/8.0/en/mathematical-functions.html#function_crc32
- https://mariadb.com/kb/en/crc32/

[the remark this answers is not in the document]
Comment (rob006) [reply framework/mutex/MysqlMutex.php:72]: I don't really trust `CRC32` in this case. :P I would rather trim lock name - `GET_LOCK(SUBSTRING(CONCAT(:prefix, :name), 0, 64), :timeout)` - lock name should be more readable and also good enough for most cases. For edge cases when you have 64 chars long database name you may need to set prefix manually.

Or just hash at DB level `GET_LOCK(SHA1(CONCAT(:prefix, :name)), :timeout)`

[the remark this answers is not in the document]
Comment (kamarton) [reply framework/mutex/MysqlMutex.php:72]: https://github.com/yiisoft/yii2/pull/19614

**Comment by rhertogh** [comment]
This PR is not backwards compatible. If an application would, for example, use different databases for each of it's customers a.k.a. tenants, the mutex would no longer work on the MySQL server level after the implementation of this PR.

I think the assumption made in #19316 is incorrect. Different databases should not matter for the mutex, it's the driver for the mutex that determines its scope. So as long as the same endpoint is used the application should be able to count on the mutex behavior of that driver (in this case the MySQL server).

If an application needs 2 different scopes for different mutexes as in #19316 the `keyPrefix` could be configured.
The current implementation forces the database name into the key (https://github.com/yiisoft/yii2/pull/19603/files#diff-4c57428f9b73c50ff9f17f5f236aac5447feddfbbd34c6dc467ba5d8e06583c4R55-R57) therefore there is no longer any way to rely on the mutex for any "global" purposes, unless the `keyPrefix` would be set in those cases (which would make it backwards incompatible).

**Comment by bizley** [comment]
I'm not sure if this reasoning is right, are you talking about different connections + different dbs or one connections + different dbs because the one case should be fine here. Or should not?

**Comment by rhertogh** [comment]
@bizley As far as I know both the connection or the database should not make any difference.

By default the "scope" of the mutex is "server wide" since the MySQL server itself stores the lock in the "performance_schema" database in the "metadata_locks" table (https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html). Therefore any client, no matter the database (you can even open up a CLI client without using any database), would request the same lock (assuming they use the same name for the lock).

So if the default behavior is a "server wide" scope for the uniqueness of a mutex lock this PR would brake that default behavior (and therefore braking backwards compatibility).

**Comment by rhertogh** [comment]
My suggestion would be to remove the default value for the `keyPrefix` (https://github.com/yiisoft/yii2/pull/19603/files#diff-4c57428f9b73c50ff9f17f5f236aac5447feddfbbd34c6dc467ba5d8e06583c4R55-R57).
If needed it then can be set in the configuration, e.g.:
```
'mutex1' => [
    'class' => 'yii\mutex\MysqlMutex',
    'db' => 'db1',
    'keyPrefix' => 'myPrefix1'
],
'mutex2' => [
    'class' => MysqlMutex::class,
    'db' => 'db2',
    'keyPrefix' => 'myPrefix2'
],
```

@samdark You mentioned before that the name _should_ be included (https://github.com/yiisoft/yii2/issues/19316#issuecomment-1092192243), what are your thoughts on this (please see comments above)?

**Comment by rob006** [comment]
@rhertogh The problem is that old behavior is unexpected and hard to debug. If you share DB server with others, you may not even be aware the problem, you just get random lock failures out of nowhere. So while this is a BC break, I think that new behavior is more sensible as a default. But this change indeed should be documented in UPGRADE.md.

**Comment by rhertogh** [comment]
@rob006 True, a note in UPGRADE.md could cover it.
It should also include the remark that a database name has to be specified in the database connection, a "generic" connection (without any specific database name) would otherwise fail with `ERROR 3057 (42000): Incorrect user-level lock name 'NULL'.`

**Comment by samdark** [comment]
@rhertogh do you have time for a pull request with additions to UPGRADE.

**Comment by rhertogh** [comment]
@samdark I've created a PR: #19645
During the writing I recognized another important point to consider, please see the "warning" in the UPGRADE.md (https://github.com/yiisoft/yii2/pull/19645/files#diff-2632bce37f2b1a645fe1a8909903481e3c4ca921d46cd9bf589752b4a3c7086eR78)

**Pull request merged**